src/ext.rs

```diff
@@ -1,10 +1,13 @@
-//! HTTP extensions
+//! HTTP extensions.
 
 use bytes::Bytes;
 #[cfg(feature = "http1")]
 use http::header::{HeaderName, IntoHeaderName, ValueIter};
 use http::HeaderMap;
 
+#[cfg(feature = "http2")]
+pub use h2::ext::Protocol;
```

Comment:
This makes the `h2` dependency version public. So we can't upgrade any h2 "breaking changes", which I'd rather we be able to do.

We _could_ expose a function to add `Protocol`, kind like how there's `hyper::upgrade::on(req)`. WDYT?

Reply:
> We _could_ expose a function to add `Protocol`, kind like how there's `hyper::upgrade::on(req)`. WDYT?

AFAIK that would still be a breaking change, as `Protocol` would still be exposed through the request's extensions.

Should I try duplicating `Protocol` here, and replacing it by h2's when actually passing the request to h2?

Reply:
Actually, the `h2` dependency is already public, through the error types.

Reply:
Anyway I added a commit to make `hyper::ext::Protocol` be a wrapper of `h2::ext::Protocol`.

Reply:
> Protocol would still be exposed through the request's extensions.

That's true... I have one regret around `hyper::upgrade::OnUpgrade`, that I made it the actual type put in extensions. If I had wrapped it in a private internal type, then it wouldn't be possible to depend on the extension type being in there.

We _could_ do similar in `h2`. Make the extension type actually private, and you have to call a function instead. Then in hyper, we just wrap calling that function. It's more of a dance, but perhaps stability is worth it. I'm actively pushing towards a 1.0, so this is more on my mind.

> Actually, the h2 dependency is already public, through the error types.

You mean through the `source`? That's an interesting point. It seems it's still an [open question](https://github.com/rust-lang/api-guidelines/discussions/96), whether that should be considered part of the public API. People seem to lean towards saying it's not, but it _would_ still break people at runtime if they depended on it... I could see an argument for us removing those connections as part of a hyper 1.0. Do you think we should consider that?

Reply:
I think rewrapping is fine, I actually never use `hyper::upgrade::on`, I don't like it, I prefer to reach in the extensions myself.

If you want to hide the type, you end up with weird traits like `CanUpgrade` and I don't like that at all.

For `h2`,  I think we should actually just bite the bullet and make it a public dependency, I've often wished I could reach into the `h2` parts of Hyper from Hyper's API.

Why do you want to push for 1.0 btw, any specific reason?